Reconnect protocol needs to finish cleanly

### Description

When the teacher node finishes a reconnect session it switches to next protocol on the same connection. The learner node, however, may take time before it exits reconnect and becomes ready to respond to the next protocol. The [former] teacher does not receive the expected response and closes the connection. If the learner is still in the reconnect protocol, it considers the closed connection as a fatal error and aborts reconnect (despite the fact that it has received all information).

A good indication of this error is a log message on the teacher side about finishing reconnect followed by another message about a broken connection. The learner reports a new connection from the teacher node and then may or may not fail reconnect depending on timing.

### Steps to reproduce

The bug is reproduced in performance networks with a large state.
Here is an excerpt from `swirlds.log` supporting the bug description:

### teacher
```
2023-11-04 12:45:41.369 63332    INFO  RECONNECT        <<platform-core: SyncProtocolWith6 5 to 6>> ReconnectTeacher: Starting reconnect in the role of the sender {"receiving":false,"nodeId":5,"otherNodeId":6,"round":271152} [com.swirlds.logging.payloads.ReconnectStartPayload]
...
2023-11-04 12:56:36.636 64468    INFO  RECONNECT        <<platform-core: SyncProtocolWith6 5 to 6>> ReconnectTeacher: Finished reconnect in the role of the sender. {"receiving":false,"nodeId":5,"otherNodeId":6,"round":271152,"success":false} [ReconnectFinishPayload]
2023-11-04 12:56:41.641 64478    WARN  SOCKET_EXCEPTIONS <<platform-core: SyncProtocolWith6 5 to 6>> NetworkUtils: Connection broken: 5 -> 6 Caused By: {Exception: SocketTimeoutException Message: Read timed out }
```

### learner
```
2023-11-04 12:45:41.382 1399     INFO  RECONNECT        <<reconnect: reconnect-controller>> ReconnectHelper: Starting reconnect in role of the receiver. {"receiving":true,"nodeId":6,"otherNodeId":5,"round":268231} [com.swirlds.logging.payloads.ReconnectStartPayload]
...
2023-11-04 12:56:12.299 1633     INFO  RECONNECT        <<reconnect: reconnect-controller>> LearningSynchronizer: receiving tree rooted at com.swirlds.virtualmap.internal.merkle.VirtualRootNode with route [0 -> 11 -> 1]
2023-11-04 12:56:41.699 1643     WARN  SOCKET_EXCEPTIONS <<sync_server #4>> InboundConnectionManager: 6 got new connection from 5, disconnecting old one
2023-11-04 12:56:42.010 1644     ERROR EXCEPTION        <<reconnect: reconnect-controller>> StandardWorkGroup: Work Group Exception [ groupName = learning-synchronizer ]
java.util.concurrent.ExecutionException: com.swirlds.common.merkle.synchronization.utility.MerkleSynchronizationException: Failed to deserialize object with class ID -7441057967105127907(0x98BC0D340D9BCA1D) (class com.swirlds.common.merkle.synchronization.internal.Lesson)
Caused by: java.io.IOException: Stream closed
```

### Additional context

See also #9678

### Hedera network

other

### Version

v0.43.0

### Operating system

Linux

--- a/platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/ByteConstants.java
+++ b/platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/ByteConstants.java
@@ -39,12 +39,6 @@ public final class ByteConstants {
     public static final byte COMM_EVENT_ABORT = 0x49 /* 73 */;
     /** sent after all events have been sent for this sync */
     public static final byte COMM_EVENT_DONE = 0x4a /* 74 */;
-    /** sent if a node wants to get the latest signed state */
-    public static final byte COMM_STATE_REQUEST = 0x4c /* 76 */;
-    /** sent as a reply to COMM_STATE_REQUEST when accepting to transfer the latest state */
-    public static final byte COMM_STATE_ACK = 0x4d /* 77 */;
-    /** sent as a reply to COMM_STATE_REQUEST when NOT accepting to transfer the latest state */
-    public static final byte COMM_STATE_NACK = 0x4e /* 78 */;
     /**
      * Private constructor to never instantiate this class
      */

--- a/platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/reconnect/ReconnectLearner.java
+++ b/platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/reconnect/ReconnectLearner.java
@@ -156,6 +156,7 @@ public class ReconnectLearner {
             receiveSignatures();
             final ReservedSignedState reservedSignedState = reconnect();
             validator.validate(reservedSignedState.get(), addressBook, stateValidationData);
+            ReconnectUtils.endReconnectHandshake(connection);
             return reservedSignedState;
         } catch (final IOException | SignedStateInvalidException e) {
             throw new ReconnectException(e);

--- a/platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/reconnect/ReconnectTeacher.java
+++ b/platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/reconnect/ReconnectTeacher.java
@@ -171,6 +171,7 @@ public class ReconnectTeacher {
         try {
             sendSignatures(signedState);
             reconnect(signedState);
+            ReconnectUtils.endReconnectHandshake(connection);
         } catch (final InterruptedException e) {
             Thread.currentThread().interrupt();
             throw new ReconnectException(e);

--- a/platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/reconnect/ReconnectUtils.java
+++ b/platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/reconnect/ReconnectUtils.java
@@ -17,17 +17,12 @@
 package com.swirlds.platform.reconnect;
 
 import static com.swirlds.logging.legacy.LogMarker.EXCEPTION;
-import static com.swirlds.logging.legacy.LogMarker.RECONNECT;
 
-import com.swirlds.common.io.exceptions.BadIOException;
 import com.swirlds.common.merkle.crypto.MerkleCryptoFactory;
-import com.swirlds.common.system.NodeId;
 import com.swirlds.logging.legacy.payload.ReconnectFailurePayload;
-import com.swirlds.platform.gossip.sync.SyncInputStream;
-import com.swirlds.platform.gossip.sync.SyncOutputStream;
-import com.swirlds.platform.network.ByteConstants;
 import com.swirlds.platform.network.Connection;
 import com.swirlds.platform.state.State;
+import edu.umd.cs.findbugs.annotations.NonNull;
 import java.io.IOException;
 import java.util.concurrent.ExecutionException;
 import org.apache.logging.log4j.LogManager;
@@ -38,72 +33,30 @@ import org.apache.logging.log4j.Logger;
  */
 public final class ReconnectUtils {
     private static final Logger logger = LogManager.getLogger(ReconnectUtils.class);
+    /**
+     * A value to send to signify the end of a reconnect. A random long value is chosen to minimize the possibility that
+     * the stream is misaligned
+     */
+    private static final long END_RECONNECT_MSG = 0x7747b5bd49693b61L;
 
     private ReconnectUtils() {}
 
     /**
-     * Validate that the other node is willing to reconnect with us.
+     * Send and receive the end reconnect message
      *
-     * @param connection
-     * 		the connection to use
-     * @return true if the other node is willing to assist with reconnect
-     * @throws IOException
-     * 		thrown when any I/O related errors occur
-     * @throws ReconnectException
-     * 		thrown when the other node is unwilling to reconnect right now
+     * @param connection the connection to send/receive on
+     * @throws IOException if the connection breaks, times out, or the wrong message is received
      */
-    public static boolean isNodeReadyForReconnect(final Connection connection) throws IOException, ReconnectException {
-        final NodeId otherId = connection.getOtherId();
-        final SyncInputStream dis = connection.getDis();
-        final SyncOutputStream dos = connection.getDos();
-
-        // send the request
-        dos.write(ByteConstants.COMM_STATE_REQUEST);
-        dos.flush();
-        logger.info(RECONNECT.getMarker(), "Requesting to reconnect with node {}.", otherId);
-
-        // read the response
-        final byte stateResponse = dis.readByte();
-        if (stateResponse == ByteConstants.COMM_STATE_ACK) {
-            logger.info(RECONNECT.getMarker(), "Node {} is willing to help this node to reconnect.", otherId);
-            return true;
-        } else if (stateResponse == ByteConstants.COMM_STATE_NACK) {
-            logger.info(
-                    RECONNECT.getMarker(),
-                    new ReconnectFailurePayload(
-                            "Node " + otherId + " is unwilling to help this node to reconnect.",
-                            ReconnectFailurePayload.CauseOfFailure.REJECTION));
-            return false;
-        } else {
-            throw new BadIOException("COMM_STATE_REQUEST was sent but reply was " + stateResponse
-                    + " instead of COMM_STATE_ACK or COMM_STATE_NACK");
+    static void endReconnectHandshake(@NonNull final Connection connection) throws IOException {
+        connection.getDos().writeLong(END_RECONNECT_MSG);
+        connection.getDos().flush();
+        final long endReconnectMsg = connection.getDis().readLong();
+        if (endReconnectMsg != END_RECONNECT_MSG) {
+            throw new IOException(
+                    "Did not receive expected end reconnect message. Expecting %x, Received %x"
+                            .formatted(END_RECONNECT_MSG, endReconnectMsg)
+            );
         }
-    }
-
-    /**
-     * Write a flag to the stream. Informs the receiver that reconnect will proceed.
-     *
-     * @param connection
-     * 		the connection to use
-     * @throws IOException
-     * 		thrown when any I/O related errors occur
-     */
-    static void confirmReconnect(final Connection connection) throws IOException {
-        connection.getDos().write(ByteConstants.COMM_STATE_ACK);
-        connection.getDos().flush();
-    }
-
-    /**
-     * Write a flag to the stream. Informs the receiver that reconnect will not proceed.
-     *
-     * @param connection
-     * 		the connection to use
-     * @throws IOException
-     * 		thrown when any I/O related errors occur
-     */
-    static void denyReconnect(final Connection connection) throws IOException {
-        connection.getDos().write(ByteConstants.COMM_STATE_NACK);
-        connection.getDos().flush();
     }
 
     /**

--- a/platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/reconnect/ReconnectUtils.java
+++ b/platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/reconnect/ReconnectUtils.java
@@ -52,10 +52,8 @@ public final class ReconnectUtils {
         connection.getDos().flush();
         final long endReconnectMsg = connection.getDis().readLong();
         if (endReconnectMsg != END_RECONNECT_MSG) {
-            throw new IOException(
-                    "Did not receive expected end reconnect message. Expecting %x, Received %x"
-                            .formatted(END_RECONNECT_MSG, endReconnectMsg)
-            );
+            throw new IOException("Did not receive expected end reconnect message. Expecting %x, Received %x"
+                    .formatted(END_RECONNECT_MSG, endReconnectMsg));
         }
     }
 

